mindmeld/components/entity_resolver.py

```diff
@@ -542,43 +231,120 @@ def _process_entity_map(entity_type,
 
     def _load_entity_map(self, force_reload=False):
-        return self._resource_loader.get_entity_map(self.type, force_reload=force_reload)
+        try:
+            return self._resource_loader.get_entity_map(self.type, force_reload=force_reload)
+        except Exception as e:
+            msg = f"Unable to load entity mapping data for " \
+                  f"entity type: {self.type} in app_path: {self.app_path}"
+            raise Exception(msg) from e
 
-    @abstractmethod
-    def _fit(self, clean, entity_map):
-        """Fits the entity resolver model
+    @staticmethod
+    def _format_entity_map(entity_map):
+        if not entity_map.get("entities", []):
+            return entity_map
+
+        # format (if required) entities key-field
+        new_entities = []
+        id_counter = 0
+        for ent in entity_map.get("entities", []):
+            cname = ent.get("cname", None)
+            if not cname:
+                continue
+            _id, whitelist = ent.get("id", None), ent.get("whitelist", [])
+            if not _id:
+                _id = str(id_counter)
+                id_counter += 1
+            new_entities.append({"id": _id, "cname": cname, "whitelist": whitelist})
+
+        entity_map["entities"] = new_entities
+        return entity_map
+
+    def _trim_results(self, results, top_n):
+        """
+        Trims down the results generated by any ER class, finally populating at max top_n documents
 
         Args:
-            clean (bool): If ``True``, deletes and recreates the index from scratch instead of
-                            updating the existing index with synonyms in the mapping.json.
-            entity_map (json): json data loaded from `mapping.json` file for the entity type
+            results (list[dict]): Each element in this list is a result dictions with keys such as
+                `id` (optional), `cname`, `score` and any others
+            top_n (int): Number of top documents required to be populated
+
+        Returns:
+            list[dict]: if trimmed, a list similar to `results` but with fewer elements,
+                        else, the `results` list as-is is returned
         """
-        raise NotImplementedError
 
-    def fit(self, clean=False):
+        if not results:
+            return None
+
+        if not isinstance(top_n, int) or top_n <= 0:
+            msg = f"The value of `top_n` set to `{top_n}` during predictions in " \
+                  f"{self.__class__.__name__}. This will result in an unsorted list of documents. "
+            logger.warning(msg)
+            return results
+
+        # Obtain top scored for each doc id, if scores field exist in results
+        best_results = {}
+        for result in results:
+            if "score" not in result:
+                return results
+            _id = result["id"] if "id" in result else result["cname"]
+            if _id not in best_results or result["score"] > best_results[_id]["score"]:
```

Comment:
its not exactly clear why `_id not in best_results` returning `True` would lead to `best_results[_id] = result`

Reply:
When `_id not in best_results`, it means that `_id` is seen for the first time in the for-loop and hence, a field with that `_id` does not already exist in  `best_results`, thereby leading to KeyError if we directly do `best_results[_id] = result`.